giraf/binaryGiraf/binReader.go

```diff
@@ -0,0 +1,269 @@
+package binaryGiraf
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"github.com/biogo/hts/bgzf"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/common"
+	"github.com/vertgenlab/gonomics/dna"
+	"github.com/vertgenlab/gonomics/dnaThreeBit"
+	"github.com/vertgenlab/gonomics/fileio"
+	"github.com/vertgenlab/gonomics/giraf"
+	"github.com/vertgenlab/gonomics/simpleGraph"
+	"io"
+	"log"
+	"strings"
+)
+
+// The BinReader struct wraps the bgzf reader from the biogo repository with a bytes buffer to store encoded giraf records
+type BinReader struct {
+	bg       *bgzf.Reader
+	currData *bytes.Buffer
+}
+
+// NewBinReader creates a new BinReader
+func NewBinReader(file io.Reader) *BinReader {
+	reader, err := bgzf.NewReader(file, 1) //TODO: Play with different levels of concurrency
+	common.ExitIfError(err)
+	return &BinReader{
+		bg:       reader,
+		currData: &bytes.Buffer{},
+	}
+}
+
+// DecompressGiraf will decode a binary giraf file (.giraf.fe) and output a giraf file (.giraf)
+func DecompressGiraf(filename string, graph *simpleGraph.SimpleGraph) {
+	// Initialize infile
+	infile := fileio.EasyOpen(filename)
+	defer infile.Close()
+	reader := NewBinReader(infile.BuffReader)
+	var err error
+
+	// Initialize outfile
+	outfile := fileio.EasyCreate(strings.TrimSuffix(filename, ".fe"))
+	defer outfile.Close()
+
+	// Read info until EOF
+	var curr giraf.Giraf
+	for curr, err = reader.Read(graph); err != io.EOF; curr, err = reader.Read(graph) {
+		common.ExitIfError(err)
+		giraf.WriteGirafToFileHandle(outfile, &curr)
+	}
+
+	// Close reader
+	err = reader.bg.Close()
+	common.ExitIfError(err)
+}
+
+// The Read method for the BinWriter struct decompresses a single giraf record and writes to file
```

Comment:
Could you tell me a bit about this syntax with the type before the func name?

Reply:
This is the syntax for declaring type methods. i.e. the Read function is a method 'owned' by the BinWriter type. This style was based on the bgzf writers from biogo that I was looking at. Most readers and writers in the standard go library seem to declare read and write as methods, but I don't see why we couldn't move this to our standard procedural syntax if that is what people want.

Reply:
Good point. Since this function is not reading bytes (and therefore won't satisfy the io.Reader interface) it would be best to stick to procedural syntax. I will make this change.